File: go-solutions/443_string_compression.go
package leetcode

import "strconv"

func Compress(chars []byte) int {
	i, res := 0, 0
	n := len(chars)
	for i < n {
		groupLen := 1
		for i+groupLen < n && chars[i] == chars[i+groupLen] {
			groupLen++
		}

		chars[res] = chars[i]
		res++

		if groupLen > 1 {
			groupLenAscii := strconv.Itoa(groupLen)
			for i := 0; i < len(groupLenAscii); i++ {
				chars[res] = groupLenAscii[i]
				res++
			}
		}
		i += groupLen
	}
	return res
}

File: ts-solutions/src/443-string-compression.ts
export function compress(chars: string[]) {
	let i = 0
	let res = 0
	while (i < chars.length) {
		let groupLen = 0
		while (i + groupLen < chars.length && chars[i] === chars[i + groupLen]) {
			groupLen++
		}
		chars[res] = chars[i]
		res++
		if (groupLen > 1) {
			const gs = groupLen.toString()
			for (const g of gs) {
				chars[res] = g
				res++
			}
		}
		i += groupLen
	}
	return res
}
